timelock/src/ibe/fullident.rs

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2024 by Ideal Labs, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use super::utils::{cross_product_32, h2, h3, h4};
+use ark_ec::Group;
+use ark_ff::{UniformRand, Zero};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::{ops::Mul, rand::Rng, vec::Vec};
+use serde::{Deserialize, Serialize};
+
+use w3f_bls::{EngineBLS, Message};
+
+/// Represents a ciphertext in the BF-IBE FullIdent scheme
+#[derive(
+	Debug,
+	Clone,
+	PartialEq,
+	CanonicalDeserialize,
+	CanonicalSerialize,
+	Serialize,
+	Deserialize,
+)]
+pub struct Ciphertext<E: EngineBLS> {
+	/// U = rP
+	pub u: E::PublicKeyGroup,
+	/// V = sigma (+) H_2(g_id^r)
+	pub v: Vec<u8>,
+	/// W = message (+) H_4(sigma)
+	pub w: Vec<u8>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum IbeError {
+	DecryptionFailed,
+}
+
+/// A type to represent an IBE identity (for which we will encrypt message)
+#[derive(Debug, Clone)]
+pub struct Identity(pub Vec<Message>);
+
+impl Identity {
+	/// construct a new identity from a string
+	pub fn new(ctx: &[u8], identities: Vec<Vec<u8>>) -> Self {
+		Self(
+			identities
+				.iter()
+				.map(|identity| Message::new(ctx, &identity))
```

Comment:
The reference isn't necessary
```suggestion
				.map(|identity| Message::new(ctx, identity))
```